arangod/RocksDBEngine/RocksDBDumpManager.cpp

```diff
@@ -0,0 +1,175 @@
+             ServerState::instance()->isDBServer());
+
+  // generating the dump context can throw exceptions. if it does, then
+  // no harm is done, and no resources will be leaked.
+  auto context = std::make_shared<RocksDBDumpContext>(
+      _engine, _engine.server().getFeature<DatabaseFeature>(), generateId(),
+      std::move(opts), user, database);
+
+  std::lock_guard mutexLocker{_lock};
+
+  if (_engine.server().isStopping()) {
+    // do not accept any further contexts when we are already shutting down
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_SHUTTING_DOWN);
+  }
+
+  bool inserted = _contexts.try_emplace(context->id(), context).second;
+  if (!inserted) {
+    // cannot insert into map. should never happen
+    TRI_ASSERT(false);
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL,
+                                   "unable to insert dump context");
+  }
+
+  TRI_ASSERT(_contexts.at(context->id()) != nullptr);
+
+  return RocksDBDumpContextGuard(*this, std::move(context));
+}
+
+RocksDBDumpContextGuard RocksDBDumpManager::find(std::string const& id,
+                                                 std::string const& database,
+                                                 std::string const& user) {
+  std::lock_guard mutexLocker{_lock};
+
+  auto it = _contexts.find(id);
+  if (it == _contexts.end()) {
+    // "cursor not found" is not a great return code, but it is much more
+    // specific than a generic error. we can also think of a dump context
+    // as a collection of cursors for shard dumping.
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_CURSOR_NOT_FOUND,
+                                   "requested dump context not found");
+  }
+
+  auto context = (*it).second;
+  TRI_ASSERT(context != nullptr);
+  if (!context->canAccess(database, user)) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_FORBIDDEN,
+                                   "insufficient permissions");
+  }
+
+  // whenever we access a context, we proactively extend its lifetime,
+  // so it does not expire quickly after
+  context->extendLifetime();
+
+  return RocksDBDumpContextGuard(*this, std::move(context));
+}
+
+void RocksDBDumpManager::remove(std::string const& id,
+                                std::string const& database,
+                                std::string const& user) {
+  std::lock_guard mutexLocker{_lock};
+
+  auto it = _contexts.find(id);
+  if (it == _contexts.end()) {
+    // "cursor not found" is not a great return code, but it is much more
+    // specific than a generic error. we can also think of a dump context
+    // as a collection of cursors for shard dumping.
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_CURSOR_NOT_FOUND,
+                                   "requested dump context not found");
+  }
+
+  auto& context = (*it).second;
+  if (!context->canAccess(database, user)) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_FORBIDDEN,
+                                   "insufficient permissions");
+  }
```

Comment:
This is a duplication of what we have in `find`. Can we factor this out into a separate function?

Reply:
Addressed.